### DISCUSSION TITLE
Saving historical data? Replay?

### DISCUSSION THREAD

**Discussion by nielsbom** [ClementTsang/bottom#1003] [Q&A]
When my machine (macOS) hangs I _want_ to see which processes are using CPU but I can't start `btm` to do so, because it's hanging 😄 

I'd like to keep it (or another utility) running to keep historical data on system resource usage and to be able to replay or view historical data.

I can imagine this being out of scope for `btm` itself, but maybe not.

Are there tools that can do this? Can `btm` potentially do this?

Thanks!

**Comment by ClementTsang** [comment]
Hm.

Off the top of my head, I don't know any utilities at a terminal level similar to bottom/htop/top/etc. that do so (wouldn't be surprised if there are any though!). I can imagine more complicated applications (e.g. monitoring tools like Prometheus + an interface of choice, etc.) easily being able to do so, but those are a bit bigger of programs and understandably might not be ideal on a personal use system.

As for whether bottom could do this - I won't say make any promises, but the idea is certainly interesting, and I can see some ways of it working.

**Reply by ClementTsang** [reply]
Just some more rambling/ideas/thoughts:

- It wouldn't be too hard to write something that optionally exports data to disk via some DB (probably SQLite) as it is processed. I'm not sure what the best way of logging processes is though other than just... dumping them. At least with time-based data there are tricks to reduce size. Or maybe this won't be a concern.
- Reading from this wouldn't be too hard either. 
- Assuming that's dealt with, I'm not sure what the interface to "rewind" would be, or at least, how to incorporate it into bottom. At least with time graphs, there's already an interface for moving back in time, even if you can't change the right-side time (yet), but doing it for tabular data like processes might be trickier.
